### DISCUSSION TITLE
Timeout on asyncio.core._io_queue.queue_read(stream)

### DISCUSSION THREAD

**Discussion by edurohr** [micropython/micropython#10894] [Using MicroPython]
Hi,

I have a function that waits asynchronously for some data to be available for reading in a stream.
I would like to be able to timeout if I don't receive the expected amount of data within a certain time.
How can this be achieved with micropython?

Original code:
```
async def read(self, bytes_to_read):
    while bytes_to_read:                
        yield asyncio.core._io_queue.queue_read(self.stream)
        data = self.stream.read(bytes_to_read)
        bytes_to_read -= len(data)
        self.in_buf += data
    
    data_to_return = self.in_buf
    self.in_buf = b""
    return data_to_return
```

Desired behaviour (with code that doesn't work):
```
async def read(self, bytes_to_read):
    while bytes_to_read:
        try:
            yield asyncio.core._io_queue.queue_read(self.stream, timeout=1_000)
        except asyncio.TimeoutError:
            self.in_buf = b""
            return            
        data = self.stream.read(bytes_to_read)
        bytes_to_read -= len(data)
        self.in_buf += data
    
    data_to_return = self.in_buf
    self.in_buf = b""
    return data_to_return
```

**Comment by peterhinch** [comment]
You need to use `wait_for`: see [this tutorial section](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#522-tasks-with-timeouts).

**Reply by edurohr** [reply]
Thanks Peter. 

I could do something like the code below, but I was hoping to get something without using too much CPU in a loop. Any ideas?
Btw, the stream in question is the USB_VCP. I don't think it has any native awaitables.

```
async def wait_for_data(stream):
    while True:
        if stream.any():
            return
        asyncio.sleep(0)

async def read(self, bytes_to_read):
    while bytes_to_read:
        try:
            yield asyncio.wait_for_ms(wait_for_data(), timeout=1_000)
        except asyncio.TimeoutError:
            self.in_buf = b""
            return            
        data = self.stream.read(bytes_to_read)
        bytes_to_read -= len(data)
        self.in_buf += data
    
    data_to_return = self.in_buf
    self.in_buf = b""
    return data_to_return
```

**Comment by peterhinch** [comment]
I have no experience of using USB VCP with `uasyncio` so I can only offer general hints.

The normal way to handle stream input is with a [StreamReader object](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#63-using-the-stream-mechanism). This uses a more efficient way to wait on I/O compared with the above busy-wait loop. The doc referenced explains how this works internally using `select.poll`.

I'm puzzled by
```python
            yield asyncio.wait_for_ms(wait_for_data(), timeout=1_000)
```
The normal way is
```python
            await asyncio.wait_for_ms(wait_for_data(), timeout=1_000)
```
